test/nightly_test.go

```diff
@@ -0,0 +1,297 @@
+// +build nightly
+
+package test
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"sort"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	testcluster "github.com/mesosphere/ksphere-testing-framework/pkg/cluster"
+	testharness "github.com/mesosphere/ksphere-testing-framework/pkg/harness"
+	"gopkg.in/yaml.v2"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	typedbatchv1 "k8s.io/client-go/kubernetes/typed/batch/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	updateFixturesEnvVar = "UPDATE_FIXTURES"
+
+	kubeaddonsNamespace                     = "kubeaddons"
+	prometheusMetricTestsFilename           = "testdata/prometheus-metric-tests.yaml"
+	testNightlyGroupPromMetricTestsFilename = "testdata/test-nightly-group-prom-metric-tests.yaml"
+)
+
+func updateFixtures() bool {
+	return os.Getenv(updateFixturesEnvVar) == "true"
+}
+
+func updateFixturesFatalMessage(err error) string {
+	return fmt.Sprintf("failing, as "+updateFixturesEnvVar+"=true: %v", err)
+}
+
+func TestNightlyGroup(t *testing.T) {
+	if err := testgroup(t, allAWSGroupName, defaultKindestNodeImage, nightlyPrometheusMetricsChecker); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// TestUnmarshallPrometheusMetricNames is a quick unit test for unmarshallPrometheusMetrics
+func TestUnmarshallPrometheusMetricNames(t *testing.T) {
+	b, err := ioutil.ReadFile("testdata/prometheus-metrics-checker.json")
+	if err != nil {
+		t.Fatal(err)
+	}
+	metrics, err := unmarshallPrometheusMetrics(b)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if updateFixtures() {
+		err := updatePrometheusMetricFixtures(prometheusMetricTestsFilename, metrics)
+		t.Fatal(updateFixturesFatalMessage(err))
+	}
+	err = testMetricFixtures(prometheusMetricTestsFilename, metrics)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func nightlyPrometheusMetricsChecker(t *testing.T, cluster testcluster.Cluster) testharness.Job {
+	return func(t *testing.T) error {
+		promMetricsBytes, err := getPromtheusMetricsBytes(t, cluster)
+		if err != nil {
+			return err
+		}
+
+		metrics, err := unmarshallPrometheusMetrics(promMetricsBytes)
+		if err != nil {
+			return err
+		}
+
+		if updateFixtures() {
+			err := updatePrometheusMetricFixtures(testNightlyGroupPromMetricTestsFilename, metrics)
+			return fmt.Errorf(updateFixturesFatalMessage(err))
+		}
+
+		err = testMetricFixtures(prometheusMetricTestsFilename, metrics)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func getPromtheusMetricsBytes(t *testing.T, cluster testcluster.Cluster) ([]byte, error) {
```

Comment:
I think this function could be simplified by port-forwarding to the prometheus server rather than running a job to query it. The ksphere-testing-framework lib makes that pretty easy, here's an example of a test where we're taking advantage of it: https://github.com/mesosphere/kubernetes-base-addons/blob/a1439fa352c85a2e8215a633cd4ce8c570f2ddcb/test/prometheus_test.go#L29 WDYT?

Reply:
ah, wasn't aware, i'll swap to that. thanks!

Reply:
changed! much better, here's the successful test run: https://teamcity.mesosphere.io/viewLog.html?tab=buildLog&buildTypeId=kubeaddons_Nightly_KubernetesBaseAddonsNightlt&buildId=3025198